### PR TITLE
Feature/chat input performance

### PR DESCRIPTION
**Pull request by flexsurfer**
Used local state and native props for input toolbar, enable mentions only when "@" is entered , unfortunately, mentions are very complex, and we still dispatch all events on every char :(

### CODE REVIEW

File: src/status_im/ui/screens/chat/components/input.cljs [hunk 1]
```diff
@@ -107,28 +107,66 @@
...
+
+(defn on-text-change [val chat-id]
+  (swap! input-texts assoc chat-id val)
+  ;;we still store it in app-db for mentions, we don't have reactions in views
+  (re-frame/dispatch [:chat.ui/set-chat-input-text val]))
+
+(defn on-change [last-text-change timeout-id mentionable-users refs chat-id sending-image args]
+  (let [text (.-text ^js (.-nativeEvent ^js args))
+        prev-text (get @input-texts chat-id)]
+    (println sending-image (and (seq prev-text) (empty? text) (not sending-image)))
```
Comment (shivekkhurana) [inline src/status_im/ui/screens/chat/components/input.cljs]: Rogue `println`

File: src/status_im/ui/screens/chat/components/input.cljs [hunk 2]
```diff
@@ -107,28 +107,66 @@
...
+(defn hide-send [{:keys [actions-ref send-ref sticker-ref]}]
+  (quo.react/set-native-props actions-ref #js {:width nil :left nil})
+  (quo.react/set-native-props send-ref #js {:width 0 :right -100})
+  (quo.react/set-native-props sticker-ref #js {:width nil :right nil}))
+
+(defn clear-input [chat-id refs]
+  (swap! input-texts dissoc chat-id)
+  (swap! mentions-enabled dissoc chat-id)
+  (hide-send refs)
+  (some-> ^js (react/current-ref (:text-input-ref refs)) .clear))
```
Comment (shivekkhurana) [inline src/status_im/ui/screens/chat/components/input.cljs]: Very cool use of `some->` 👏

File: src/status_im/ui/screens/chat/toolbar_content.cljs [hunk 3]
```diff
@@ -31,15 +31,9 @@
...
-                                    color
-                                    chat-id
-                                    contacts
-                                    chat-type
-                                    chat-name
-                                    public?]}]
-  (when chat-id
+(defn toolbar-content-view []
+  (let [{:keys [group-chat invitation-admin color chat-id contacts chat-type chat-name public?]}
+        @(re-frame/subscribe [:chats/current-chat])]
```
Comment (shivekkhurana) [inline src/status_im/ui/screens/chat/toolbar_content.cljs:36]: Did you miss a `(when chat-id` here or is it not required anymore?

File: src/status_im/ui/screens/chat/views.cljs [hunk 4]
```diff
@@ -284,77 +311,65 @@
...
-       ;;load and render 20 messages more, but we can't prevent this , because otherwise :on-end-reached will work wrong
-       :on-end-reached               (fn []
-                                       (if @state/scrolling
-                                         (re-frame/dispatch [:chat.ui/load-more-messages chat-id])
-                                         (utils/set-timeout #(re-frame/dispatch [:chat.ui/load-more-messages chat-id])
-                                                            (if platform/low-device? 500 200))))
-
-       :on-scroll-to-index-failed    #()                    ;;don't remove this
+       :on-end-reached               list-on-end-reached
+       :on-scroll-to-index-failed    identity              ;;don't remove this
```
Comment (shivekkhurana) [inline src/status_im/ui/screens/chat/views.cljs:310]: Can you add some more context about the `identity`?

File: src/status_im/ui/screens/wallet/send/views.cljs [hunk 5]
```diff
@@ -108,39 +108,33 @@
...
-         :right
-         [quo/button
-          {:accessibility-label :select-account-bottom-sheet
-           :disabled            (nil? from)
-           :on-press            #(re-frame/dispatch
-                                  [::commands/accept-request-address-for-transaction
-                                   (:message-id message)
-                                   (:address from)])}
-          (i18n/label :t/share)]}]])))
+(defn select-account-sheet [{:keys [from message]}]
```
Comment (shivekkhurana) [inline src/status_im/ui/screens/wallet/send/views.cljs:111]: Why did we go from `views/defview` to `defn`? Is there a performance cost for using `defview` macro?

**Review by shivekkhurana — CHANGES_REQUESTED** [review]
Heavy stuff, hope this improves chat performance. Left some comments :)

File: src/status_im/ui/screens/chat/components/input.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/components/input.cljs]: yeah i also like it, thanks to someone :)

File: src/status_im/ui/screens/chat/toolbar_content.cljs (hunk 3 above)
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/toolbar_content.cljs:36]: no, i don't need it anymore

File: src/status_im/ui/screens/chat/views.cljs (hunk 4 above)
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/views.cljs:310]: I just changed anonymous empty function on identity

File: src/status_im/ui/screens/wallet/send/views.cljs (hunk 5 above)
Comment (flexsurfer) [reply src/status_im/ui/screens/wallet/send/views.cljs:111]: there are no subscriptions anymore, usually, we use defview for subscriptions

**Comment by flexsurfer** [comment]
@shivekkhurana , thanks for the review, i've addressed all your points

**Review by shivekkhurana — APPROVED** [review]
Looks good, needs QA.

**Comment by churik** [comment]
@flexsurfer
### ISSUE 1: No background colour on buttons (Android only)
photo_2021-03-23 14 39 51 jpeg 2021-03-23 14-43-04

### ISSUE 2: White line in dark mode under chat input (Android only)
photo_2021-03-23 14 40 02 jpeg 2021-03-23 14-45-59

**Comment by churik** [comment]
### ISSUE 3: Several avatars are shown for the sender if receive several messages on chat view

**Steps:**
1) user A: restore multiaccount from seed phrase
2) user B: create multiaccount, join public chat
3) user A: join same public chat
4) user A: send several messages
photo_2021-03-23 15 23 21 jpeg 2021-03-23 15-24-24
**OS:** IOS, Android

### ISSUE 4: messages are confirmed only after reopening chat view

**Steps:**
1) restore multiaccount from seed phrase
2) join public chat and send several messages

**Expected result:**
messages are confirmed immediately 

**Actual result:**
messages are confirmed only if reopening chat view

https://user-images.githubusercontent.com/4557972/112162613-7f58c980-8bec-11eb-9146-558bd305dfdb.mp4

**OS:** IOS, Android

**Comment by flexsurfer** [comment]
thanks @churik all fixed

**Comment by churik** [comment]
Reasons of failed e2e:
### ISSUE 5: No `chat-menu-button` accessibility-id on options button
Appium 2021-03-24 09-08-11

### ISSUE 6: No `back-button` accessibility-id on back button on chat view
Appium 2021-03-24 09-11-07

### ISSUE 7: when messages are loading, instead of spinner label is shown
On current nightly:
https://user-images.githubusercontent.com/4557972/112277649-3f442600-8c82-11eb-8b14-7c6a08d70942.mp4

On PR build:
https://user-images.githubusercontent.com/4557972/112277700-4c611500-8c82-11eb-8609-a14962d605b8.mp4

### ISSUE 8:  Copy and Reply options are not working (Android only)

https://user-images.githubusercontent.com/4557972/112279761-90edb000-8c84-11eb-9f30-ac065dffeb04.mp4

**Comment by flexsurfer** [comment]
thanks @churik fixed

**Comment by churik** [comment]
@flexsurfer thanks for the fixes, it works now! still testing.
One more small issue:
### ISSUE 9: Font remains blue after mentioning someone in chat (IOS only)
https://user-images.githubusercontent.com/4557972/112447703-e774f000-8d51-11eb-8659-7ba6f7ecb9da.mp4

**Comment by flexsurfer** [comment]
thanks @churik fixed

**Comment by churik** [comment]
All issues are fixed, no regression found.
Thanks a lot @flexsurfer

**Pull request merged**